src/h2_stream.erl

```diff
@@ -159,45 +183,57 @@ send_connection_window_update(Size) ->
     gen_statem:cast(self(), {send_connection_window_update, Size}).
 
 rst_stream(Pid, Code) ->
-    gen_statem:call(Pid, {rst_stream, Code}).
+    gen_statem:cast(Pid, {rst_stream, Code}).
 
 -spec stop(pid()) -> ok.
 stop(Pid) ->
     gen_statem:stop(Pid).
 
 init([
       StreamId,
+      Streams,
       ConnectionPid,
       CB=undefined,
-      _CBOptions,
-      Type,
-      Socket
+      _CBOptions
      ]) ->
+    erlang:monitor(process, ConnectionPid),
     {ok, idle, #stream_state{
                   callback_mod=CB,
-                  socket=Socket,
                   stream_id=StreamId,
+                  streams=Streams,
                   connection=ConnectionPid,
-                  type = Type
+                  socket=h2_stream_set:socket(Streams),
+                  type=h2_stream_set:stream_set_type(Streams)
                  }};
 init([
       StreamId,
+      Streams,
       ConnectionPid,
       CB,
-      CBOptions,
-      Type,
-      Socket
+      CBOptions
      ]) ->
-    %% TODO: Check for CB implementing this behaviour
-    {ok, NewCBState} = callback(CB, init, [ConnectionPid, StreamId], [Socket | CBOptions]),
+    erlang:monitor(process, ConnectionPid),
+    %% don't block stream init with a slow callback init
+    self() ! {init_callback, CBOptions},
```

Comment:
Can this not be done with the gen_statem `next_event` action that can be returned from `init`?

Reply:
yeah, I forgot about that, will fix